.github/workflows/test.yml

```diff
@@ -0,0 +1,45 @@
+name: Run Tests
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    types: ['opened', 'reopened', 'synchronize']
+
+jobs:
+  test:
+    strategy:
+      matrix:
+        nodejs: [ '18', '20', '21', '22' ]
+    runs-on: ubuntu-latest
+    container: ubuntu:latest
+    permissions:
+      actions: write
+      contents: read
+      pull-requests: write
+    services:
+      redis:
+        image: redis
```

Comment:
do you need to use a service or could you use this https://github.com/marketplace/actions/actions-setup-redis